checker/src/org/checkerframework/checker/lock/LockStore.java

```diff
@@ -136,4 +127,126 @@ protected void internalDotOutput(StringBuilder result) {
                 + "\\n");
         super.internalDotOutput(result);
     }
+
+    @Override
+    protected boolean isSideEffectFree(AnnotatedTypeFactory atypeFactory,
+            ExecutableElement method) {
+        LockAnnotatedTypeFactory lockAnnotatedTypeFactory = (LockAnnotatedTypeFactory) atypeFactory;
+        return ((LockChecker) lockAnnotatedTypeFactory.getContext()).hasOption("assumeSideEffectFree") ||
+                lockAnnotatedTypeFactory.methodSideEffectAnnotation(method, false) == SideEffectAnnotation.RELEASESNOLOCKS ||
+               super.isSideEffectFree(atypeFactory, method);
+    }
+
+    @Override
+    public void updateForMethodCall(MethodInvocationNode n,
+        AnnotatedTypeFactory atypeFactory, CFValue val) {
+        super.updateForMethodCall(n, atypeFactory, val);
+        ExecutableElement method = n.getTarget().getMethod();
+        if (!isSideEffectFree(atypeFactory, method)) {
+            // Necessary because a method could unlock a lock that is a local variable, e.g.:
+            // ReentrantLock lock = new ReentrantLock();
+            // lock.lock();
+            // unlockMyLock(lock);
+            localVariableValues.clear();
+        }
+    }
+
+    boolean hasLockHeld(CFValue value) {
+        assert value != null;
+        AnnotatedTypeMirror type = value.getType();
+        if (type != null) {
+            AnnotationMirror anno = type.getAnnotationInHierarchy(LOCKPOSSIBLYHELD);
+            if (anno != null)
+                return anno.equals(LOCKHELD);
+        }
+
+        return false;
+    }
+
+    boolean hasLockPossiblyHeld(CFValue value) {
+        assert value != null;
+        AnnotatedTypeMirror type = value.getType();
+        if (type != null) {
+            AnnotationMirror anno = type.getAnnotationInHierarchy(LOCKPOSSIBLYHELD);
+            if (anno != null)
+                return anno.equals(LOCKPOSSIBLYHELD);
+        }
+
+        return false;
+    }
+
+    @Override
+    public void insertValue(FlowExpressions.Receiver r, /*@Nullable*/ CFValue value) {
```

Comment:
This seems to be a copy of CFAbstractStore except for call hasLockHeld(...).  Instead of copying, do this:
````
if (value == null) {
  super.insertValue(...)
} else if (r instanceof FlowExpressions.FieldAccess && hasLockHeld(value)) {
  // Some comment here about why lock has to be held
  super.insertValue(...)
} else {
  super.insertValue(...)
}
````